File: GameServer/GameServerApp/GameServerApp/Proto/CreateProto/MailRequestProto.cs
//===================================================
//作    者：xxx
//创建时间：2019-05-18 15:13:23
//备    注：
//===================================================
using System.Collections;
using System.Collections.Generic;
using System;

/// <summary> 邮件(请求) </summary>
public struct MailRequestProto : IProto
{
    public ushort ProtoCode { get { return 10002; } }


    public byte[] ToArray()
    {
        using (MMO_MemoryStream ms = new MMO_MemoryStream())
        {
            ms.WriteUShort(ProtoCode);
            return ms.ToArray();
        }
    }

    public static MailRequestProto GetProto(byte[] buffer)
    {
        MailRequestProto proto = new MailRequestProto();
        using (MMO_MemoryStream ms = new MMO_MemoryStream(buffer))
        {
        }
        return proto;
    }
}

File: MMORPG_2/Assets/Script/Data/Proto/CreateLuaProto/MailRequestProto.lua
--邮件(请求)
MailRequestProto = { ProtoCode = 10002 }

--这句是重定义元表的索引，就是说有了这句，这个才是一个类
MailRequestProto.__index = MailRequestProto;

function MailRequestProto.New()
    local self = { }; --初始化self
    setmetatable(self, MailRequestProto); --将self的元表设定为Class
    return self;
end


--发送协议
function MailRequestProto.SendProto(proto)

    local ms = CS.LuaHelper.Instance:CreateMemoryStream();
    ms:WriteUShort(proto.ProtoCode);


    CS.LuaHelper.Instance:SendProto(ms:ToArray());
    ms:Dispose();
end


--解析协议
function MailRequestProto.GetProto(buffer)

    local proto = MailRequestProto.New(); --实例化一个协议对象
    local ms = CS.LuaHelper.Instance:CreateMemoryStream(buffer);


    ms:Dispose();
    return proto;
end